File: c++/count-good-nodes-in-binary-tree.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int goodNodes(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        return 1 + recursiveGoodNodes(root->left, root->val) + recursiveGoodNodes(root->right, root->val);
    }

    int recursiveGoodNodes(TreeNode* root, int largestAncestor) {
        if (root == nullptr) {
            return 0;
        }
        if (root->val < largestAncestor) {
            return recursiveGoodNodes(root->left, largestAncestor) + recursiveGoodNodes(root->right, largestAncestor);
        } else {
            return 1 + recursiveGoodNodes(root->left, root->val) + recursiveGoodNodes(root->right, root->val);
        }
    }
};

File: python/count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if not root:
            return 0

        goodCount = 1

        def recursiveGoodCheck(root: Optional[TreeNode], pathHighest: int):
            if not root:
                return
            nonlocal goodCount
            if root.val >= pathHighest:
                goodCount += 1
            curHighest = max(root.val, pathHighest)
            recursiveGoodCheck(root.left, curHighest)
            recursiveGoodCheck(root.right, curHighest)

        recursiveGoodCheck(root.left, root.val)
        recursiveGoodCheck(root.right, root.val)
        return goodCount
